src/app/modules/shared_models/message_item.nim

```diff
@@ -47,6 +48,7 @@ type
     deletedByContactDetails: ContactDetails
     links: seq[string]
     linkPreviewModel: link_preview_model.Model
+    emojiReactionsModel: emoji_reactions_model.Model
```

Comment:
So indeed we have 2 models now. 
1 is used in the "reaction popups" and the second is used to show reactions in `StatusMessage`.

Maybe we could merge them somehow?

Reply:
thanks for the comments, 

- reactionsModel is the model of reactions on a specific message, the size of this model varies depending on how many reactions are on that specific message
- emojiReactionsModel is the model of all possible reactions enhanced with what reaction is actually clicked using the didIReactWithThisEmoji role (this allow us to highlight the emoji for instance). This model has a fixed size (until we had more emojis, today we support only 6 emojis for instance)

I am currently not sure how to merge them for accomplishing those very different purposes here. I understand they share some properties.

Reply:
I'll  send a second revision with first with 2 models, and spends some more time for merging the models as I believe this is the most critical comment here.

Reply:
Soooo. Speaking of merging these models.
Let's take a look at them:

### OLD model

``` nim
type 
  Model* = ref object of QAbstractListModel
    items: seq[EmojiReactionItem]

  ModelRole {.pure.} = enum
    EmojiId = UserRole + 1
    Filename
    DidIReactWithThisEmoji
```

### NEW model

```nim
type
  MessageReactionModel* = ref object of QAbstractListModel
    items: seq[MessageReactionItem]

  ModelRole {.pure.} = enum
    EmojiId = UserRole + 1
    DidIReactWithThisEmoji
    NumberOfReactions
    JsonArrayOfUsersReactedWithThisEmoji
```

So we see common 

- `EmojiId` - common
- `DidIReactWithThisEmoji` - common
- `filename` is missing in the OLD model, because the images used by OLD model are defined in the code below. And this is a duplicated in the new code.
https://github.com/status-im/status-desktop/blob/e0fcc88934a16ff116366e24b97eecda3e6b2f98/ui/imports/shared/views/chat/MessageView.qml#L598-L605
- `NumberOfReactions` and `JsonArrayOfUsersReactedWithThisEmoji` are not needed for the UI where the NEW model is used. But it won't disturb here.

Yes, the OLD model only contains the reactions added by the used to this message. But we can simply apply a `SortFilterProxyModel` in QML to filter out what we need.

wdyt?

Reply:
Discussed offline, action is to rename the models to avoid confusing model names.

Proposal : 
- reactionsModel      --> messageReactedEmojisModel   
- emojiReactionsModel --> chatEmojiReactionsModel

Reply:
I mean we could use a single model with all possible reactions, with all required roles for both usages. 
And in one case we would filter out by `DidIReactWithThisEmoji == true`

Reply:
> I mean we could use a single model with all possible reactions, with all required roles for both usages. And in one case we would filter out by `DidIReactWithThisEmoji == true`

sorry again but still confused about your idea.

In which case we will filter out by `DidIReactWithThisEmoji == true` ? 

- if reactionsModel case : How to display the reactions by other people if I didn't react to a message?
- if emojiReactionsModel case : How the view would guess that we're missing a reactions, does not seem like the view role to check that...


I realize maybe I wasn't clear enough.

I am thinking about the add/remove reaction action which decreases the number of reactions on the message reaction model by one.
If we go that way (Having a single model for both use cases with a model filtering on the `DidIReactWithThisEmoji == true`) we need to introduce new variable(s) to be able to control above mentioned case.

For instance imagine 
M  : message, 
r* : reaction done by me
r  : reaction not done by me 

M [ 2r1*, 1r2*, 3r3 ]

2r1* means 2 people (including me) reacted with r1
1r2* means 1 person (including me) reacted with r2
3r3  means 3 people                reacted with r3

In this case, because I didn't react with r3, it'll be filtered out? it should not right? because I want to see other people reactions.

Unless I didn't understand your idea, here is the complexity I was referring to :

### Proposal

We need to use the counter attached to each reactions in a message.
The way it will work is that we will always have all the reactions available, but, in :

- case 1 : reactionsModel
   - add reaction will not add new reaction, but just increment counter
   - remove reaction will not delete reaction, but just decrement counter
   - if counter is 0, it's not displayed but not removed from the model
   - if counter  > 0 and I reacted, it's highlighted blue
   - if counter  > 0 and I did NOT react, it's highlighted grey
   - the SortFilterProxyModel will be on the number of reactions (> 0) instead

- case 2 : emojiReactionsModel
   - no SortFilterProxyModel 
   - if counter is 0, it's not highlighted
   - if counter  > 0, it's highlighted

This way, the view will adapt for the same data coming from the model. the size of the model will not vary during the execution. This requires modification in the model, and adaptation in the view for each cases.

Reply:
Sorry, your screenshot mislead me and I though for `reactionsModel` we only show our reactions 😄 

I agree with your proposal 👍 
So we keep the filtering idea, but filter based on the counter. 

The counter is actually the length of `JsonArrayOfUsersReactedWithThisEmoji`, which you will still need for hovering the reaction and showing reacted users.

-----

btw, speaking of `JsonArrayOfUsersReactedWithThisEmoji`, would be very nice to refactor this and properly pass a model of users. Probably for this "model of users" we only need a display name (or smth like that). This will allow us to subscribe to contact's changes and update the displayName in the reactions list.

It's not important, I totally leave the decision to you. Maybe open a separate issue for this.

Reply:
Ahh yes sorry for the screenshot, yeah it's a subset of the use cases...
You'll hate me, but I would love to implement squish tests even before jumping into this honestly...
How about this is merged, then I take on squish, then I do the models merging 😄
Otherwise, let me reach out to @iurimatias to let him know proactively that this ticket would take a bit more time than I had expected.